index-scheduler/src/lib.rs

```diff
@@ -407,6 +430,60 @@ impl IndexScheduler {
         Ok(this)
     }
 
+    fn index_budget(
+        tasks_path: &Path,
+        base_map_size: usize,
+        mut task_db_size: usize,
+        max_index_count: usize,
+    ) -> IndexBudget {
+        let budget = utils::dichotomic_search(base_map_size, |map_size| {
+            Self::is_good_heed(tasks_path, map_size)
+        });
+
+        log::debug!("memmap budget: {budget}B");
+        let mut budget = budget / 2;
+        if task_db_size > (budget / 2) {
+            task_db_size = clamp_to_page_size(budget * 2 / 5);
+            log::debug!(
+                "Decreasing max size of task DB to {task_db_size}B due to constrained memory space"
+            );
+        }
+        budget -= task_db_size;
+
+        // won't be mutated again
+        let budget = budget;
+        let task_db_size = task_db_size;
+
+        log::debug!("index budget: {budget}B");
+        let mut index_count = budget / base_map_size;
+        if index_count < 2 {
+            // take a bit less than half than the budget to make sure we can always afford to open an index
+            let map_size = (budget * 2) / 5;
+            // single index of max budget
+            log::debug!("1 index of {map_size}B can be opened simultaneously.");
+            return IndexBudget { map_size, index_count: 1, task_db_size };
+        }
+        // give us some space for an additional index when the cache is already full
+        // decrement is OK because index_count >= 2.
+        index_count -= 1;
+        if index_count > max_index_count {
+            index_count = max_index_count;
+        }
+        log::debug!("Up to {index_count} indexes of {base_map_size}B opened simultaneously.");
+        IndexBudget { map_size: base_map_size, index_count, task_db_size }
+    }
+
+    fn is_good_heed(tasks_path: &Path, map_size: usize) -> bool {
+        if let Ok(env) =
+            heed::EnvOpenOptions::new().map_size(clamp_to_page_size(map_size)).open(tasks_path)
+        {
```

Comment:
Unfortunately this is not easy to do.

- The dichotomic routine is not written to take into account the possibility of error, although we could change it at the cost of increased complexity
- More annoyingly, while unices are rather well-behaved and will return an OS error 12 when unable to allocate, Windows returns up to 4 different errors, making for a total of 5 cases to handle. I think it is best handled in a follow-up PR.

Looking at various scenarios that can happen here:

1. There is a persistent error wrt to opening the task db that is never resolved. In that case, we will fail opening the task db afterwards and get the error at that point.
2. There is a transient error that is resolved after computing the budget, but before opening the task db. In that case the returned budget will be 0 (1 index of 0 bytes, and the task_db_size will be computed to 0), which will attempt to open the indexes to their "natural size" (or the default size in lmdb if creating the index). This will probably render the instance unusable though as the task db will be MDB_MAP_FULL very early. A restart will be necessary.
3. There is a transient error that is resolved while computing the budget: the budget will be lower than it should be. How much of an issue this poses is dependent on when the issue is resolved. The closer we get to 0 budget, the closer to solution (2) we get.
4. There is a transient error that appear after computing the budget: in that case the budget might not be actual anymore and we'll see allocation errors pop when opening indexes. I try to mitigate that case by taking a fraction of the budget (because the budget is shared with all allocations in the application and can suffer from memory fragmentation), but ultimately this PR cannot address the possible issues caused by e.g. a different part of the programming memory mapping with a big size or making big allocations.

I'll update the PR to comment at least on the "persistent error" scenario, which looks the most likely to me.

Reply:
Thank you very much @dureuill, for this explanation. You can also link to this GitHub comment if the reader needs more info 😃

Reply:
Done, see https://github.com/meilisearch/meilisearch/pull/3331/commits/b5489fdb158e293951c6240ea452096a311dcf73